wazp/callbacks/metadata.py

```diff
@@ -519,3 +580,123 @@ def modify_rows_selection(
             alert_state,
             alert_message,
         )
+
+    @app.callback(
+        Output("import-message", "is_open"),
+        Output("import-message", "children"),
+        Output("import-message", "color"),
+        Input("upload-spreadsheet", "contents"),
+        State("upload-spreadsheet", "filename"),
+        State("import-message", "is_open"),
+        State("session-storage", "data"),
+    )
+    def generate_yaml_files_from_spreadsheet(
+        spreadsheet_uploaded_content: str,
+        spreadsheet_filename: str,
+        import_message_state: bool,
+        app_storage: dict,
+    ):
+        """Generate yaml files from spreadsheet
+
+        From example at
+        https://dash.plotly.com/dash-core-components/upload#displaying-uploaded-spreadsheet-contents
+
+        Parameters
+        ----------
+        spreadsheet_uploaded_content : str
+            _description_
+        spreadsheet_filename : str
+            _description_
+        import_message_state : bool
+            _description_
+
+        Returns
+        -------
+        _type_
+            _description_
+        """
+        import_message_text = ""
+        import_message_color = "warning"
+
+        # if data is uploaded: read uploaded content as a dataframe
+        if spreadsheet_uploaded_content is not None:
+            _, content_string = spreadsheet_uploaded_content.split(",")
+            decoded = base64.b64decode(content_string)
+            try:
+                # as csv
+                if "csv" in pl.Path(spreadsheet_filename).suffix:
+                    df = pd.read_csv(io.StringIO(decoded.decode("utf-8")))
+                # as xls(x)
+                elif "xls" in pl.Path(spreadsheet_filename).suffix:
+                    df = pd.read_excel(io.BytesIO(decoded))
+                else:
+                    import_message_state = True
+                    import_message_text = "Only csv or xls(x) files accepted"
+                    import_message_color = "warning"
+
+                    return (
+                        import_message_state,
+                        import_message_text,
+                        import_message_color,
+                    )
+
+            except Exception as e:
+                print(e)
+                import_message_state = True
+                import_message_text = (
+                    "There was an error reading" f" this file ({e})."
+                )
+                import_message_color = "danger"
+
+            # convert all fields in dataframe to strings
+            # (otherwise datetime fields are not encoded correctly in the YAML)
+            # TODO: is it better to use to_json instead?
+            # if so I think date encodes in
+            # epoch milliseconds
+            # list_dict_per_row = df.to_json('records')
+            df = df.applymap(str)
```

Comment:
So I can import your test csv that you sent me, but the test xlsx throws
```
UnboundLocalError: cannot access local variable 'df' where it is not associated with a value
```
which, I guess, is from this line.


... This might be an issue because I deleted the empty rows and saved it in [Numbers](https://www.apple.com/uk/numbers/) (not Excel). If Sam broke the xlsx file, I think you can ignore this comment.


---
Here's the file, can you reproduce?

[test_from_spreadsheet.xlsx](https://github.com/SainsburyWellcomeCentre/WAZP/files/11250679/test_from_spreadsheet.xlsx)

Reply:
Although annoying that `pd.read_excel` doesn't throw a noisier error if it fails.

Reply:
I had a look ([x]) and it seems that installing openpyxl in the environment solved the issue.... 🤔